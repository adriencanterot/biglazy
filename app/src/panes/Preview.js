import React, { Component } from 'react'
import { Classes, NonIdealState, H5, H3, H2 } from '@blueprintjs/core'
import { Cell, Table, Column, SelectionModes } from '@blueprintjs/table'
import { ProtocolLoading } from './Loading'
import {
    getProducts,
    getWrappedForProduct,
    getFormattedProducts,
    getFormattedCycles,
} from './protocolUtils'
import { connect } from 'react-redux'
/* import { createSelector } from 'reselect' */
import './Preview.css'

export const ProductLine = (protocol, product) => (
    <div className="productLine">
        <strong>{product.name}</strong>
        <br />
        {getWrappedForProduct(protocol, product).map(wrapped => (
            <span
                className={wrapped.isDelivered ? 'delivered' : 'notDelivered'}
            >
                J{wrapped.day}
            </span>
        ))}
    </div>
)

const dayList = (protocol, products) => (
    <div className="productLines">
        {products.map(product => ProductLine(protocol, product))}
    </div>
)

export const Preview = ({ protocol, products, loading }) => {
    if (loading === 'LOADING') return ProtocolLoading
    else if (loading === 'NON_IDEAL') {
        return (
            <NonIdealState
                className="nonIdealState"
                icon="search"
                title="Chimiothérapie"
                description="Selectionnez un protocole dans la liste de gauche"
            />
        )
    } else {
        return (
            <div className="Preview">
                <div className="firstLine">
                    <H2 className="protocolName">{protocol.name}</H2>
                    <div className="dayOneEquals">
                        J1 = J{protocol.dayOneEquals}
                    </div>
                </div>
                <br />
                <p>
                    Chimiothérapie comprenant <strong>{}</strong>. réévaluée
                    tous les <strong>{getFormattedCycles(protocol)}</strong>{' '}
                    cycles avec ...
                </p>
                <br />
                <div className="evaluations">
                    <H5>Réévaluations : </H5>
                    <div className="evaluationBox" key={protocol.dayOneEquals}>
                        {protocol.evaluations.map(e => (
                            <div className="flexLine" key={e.day}>
                                <div className="evaluation">
                                    {e.dayAfter} cycles
                                </div>
                                <div className="evaluationContent">
                                    {e.consultation && (
                                        <strong>Consultation</strong>
                                    )}
                                    {e.imagery && (
                                        <span>
                                            &nbsp;et <strong>Imagerie</strong>
                                        </span>
                                    )}
                                </div>
                            </div>
                        ))}
                    </div>
                    <br />
                    <br />
                    <H5>Produits : </H5>
                    {dayList(protocol, products)}
                </div>
            </div>
        )
    }
}
/*
 * When Reselect will be needed
 *
 * const getProtocol = state => state.protocol
 * const getProductsForProtocol = createSelector([getProtocol], getProducts)
 */

function mapStateToProps(state, ownProps) {
    const { loading } = state.panes
    if (loading !== 'LOADED') return { loading } // early exit

    return {
        loading: state.panes.loading,
        protocol: state.protocol,
        products: getProducts(state.protocol),
    }
}

export default connect(mapStateToProps)(Preview)
