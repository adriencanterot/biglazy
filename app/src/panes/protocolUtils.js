export const getFormattedProducts = products =>
    products.reduce((p1, p2) => p1 + ', ' + p2)
export const getFormattedCycles = protocol =>
    protocol.evaluations
        .map(e => e.dayAfter)
        .reduce((d1, d2) => d1 + (d2 && ', ' + d2))
export function getProducts(protocol) {
    const products = protocol.days.flatMap(d => d.products)
    let uniqueProducts = []
    for (let product of products) {
        if (!uniqueProducts.includes(product)) {
            uniqueProducts.push(product)
        }
    }
    return uniqueProducts
}

function parseDay(dayElement) {
    const day = dayElement.day
    if (day.includes('-')) {
        const firstDay = parseInt(day.split('-')[0], 10)
        const lastDay = parseInt(day.split('-')[1], 10)
        let returnDays = []
        for (let i = firstDay; i <= lastDay; i++) {
            returnDays.push(i)
        }
        return returnDays
    } else if (day.includes(',')) {
        const firstDay = parseInt(day.split(',')[0], 10)
        const lastDay = parseInt(day.split(',')[1], 10)
        return [firstDay, lastDay]
    } else {
        return [parseInt(day, 10)]
    }
}

export function getDaysForProduct(protocol, product) {
    let daysOfInterest = []
    for (let day of protocol.days) {
        if (day.products.includes(product)) {
            daysOfInterest.push(...parseDay(day))
        }
    }
    const sortedDaysOfInterest = daysOfInterest.sort((a, b) => a - b)
    return sortedDaysOfInterest
}

export function getWrappedForProduct(protocol, product) {
    let wrapped = []
    let deliveredDays = getDaysForProduct(protocol, product)
    for (let day = 1; day <= protocol.dayOneEquals; day++) {
        wrapped.push({
            day: day,
            isDelivered: deliveredDays.includes(day),
        })
    }
    return wrapped
}
