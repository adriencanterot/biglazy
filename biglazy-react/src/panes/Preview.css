.Preview {
    width: 80%;
}

.firstLine {
    display: flex;
    width: 100%;
    align-items: baseline;
    justify-content: space-between;
}

.firstLine > .protocolName {
    align-self: flex-start;
}

.firstLine > .dayOneEquals {
    align-self: flex-end;
}

.delivered {
    display: inline-block;
    padding: auto;
    text-align: center;
    min-width: 30px;
    padding-top: 5px;
    height: 30px;
    border-radius: 4px;
    font-weight: bold;
    background-color: #B3CFFF;
    color: #1F4B99;
    margin: 3px;
}
.notDelivered {
    display: inline-block;
    padding: auto;
    color: rgba(0, 0, 0 ,0);
    border: 1px solid grey;
    width: 24px;
    height: 24px;
    border-radius: 4px;
    margin:6px;
}
.productLine {
    padding: 5px 0px 0px 5px;
}

.dayOneEquals {
    display: inline-block;
    border-radius: 4px;
    background-color: #B3CFFF;
    font-weight: bold;
    color: #1F4B99;
    font-size: 14pt;
    text-align: center;
    padding-top: 5px;
    padding: 10px;
}

.evaluationBox {
    display: flex;
    flex-direction: column;
    border: 1px solid green;
    border-radius: 4px;
    padding: 10px;
}

.flexLine {
    display: flex;
    flex-direction: row;
    align-items: center;
}

.evaluationContent {
    margin-left: 20px;
}

.evaluation {
    padding-top: 5px;
    margin-right: 1000px;
    padding: 9px;
    margin: 5px;
    height: 40px;
    width: 15%;
    text-align: center;
    font-size: 12pt;
    border-radius: 4px;
}