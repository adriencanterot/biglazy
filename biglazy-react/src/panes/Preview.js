import React, { Component } from 'react'
import { Classes } from '@blueprintjs/core';
import { Cell, Table, Column, SelectionModes } from '@blueprintjs/table'

const contentCellRenderer = (rowIndex) => {

}

const Preview = (props) => {
    const { protocol } = props

    if(protocol.protocole != undefined) {
        const cellRenderer = (rowIndex) => <Cell>À {protocol.reevaluations[rowIndex].j} semaines</Cell>
        const imagerieCellRenderer = (rowIndex) => <Cell>{protocol.reevaluations[rowIndex].contenu.imagerie ? "Oui" : "Non" }</Cell>
        const consultationCellRenderer = (rowIndex) => <Cell>{protocol.reevaluations[rowIndex].contenu.consultation ? "Oui" : "Non" }</Cell>
        
        return (
            <div>
                <h4>{protocol.protocole}</h4>
                <p>J1 = J{protocol.dureeCycle}</p>
                <h5>Réévaluations</h5>
                <Table enableRowHeader={false} numRows={protocol.reevaluations.length} selectionModes={SelectionModes.NONE}>
                    <Column name="Reevaluations" cellRenderer={cellRenderer}/>
                    <Column name="Imagerie" cellRenderer={imagerieCellRenderer}/>
                    <Column name="Consultation" cellRenderer={consultationCellRenderer} />
                </Table>
            </div>
            
        )
    } else {
        return <div className={Classes.SKELETON}>Select a class mate</div>
    }
    
}

export default Preview