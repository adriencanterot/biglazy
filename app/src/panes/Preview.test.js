import React from 'react'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import { getProducts } from './protocolUtils'

import { Preview } from './Preview'
import { iteratee } from 'lodash'
import mockProtocol from '../test/protocol.json'

describe('Preview Component', () => {
    it('Loads and displays Preview', () => {
        const { container, asFragment } = render(
            <Preview
                protocol={mockProtocol}
                products={getProducts(mockProtocol)}
                loading={false}
            />
        )
        console.log(container)
        console.log(asFragment)
    })
})
