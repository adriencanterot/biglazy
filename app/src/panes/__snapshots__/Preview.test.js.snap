// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview Component Loads and displays Preview 1`] = `
<div>
  <div
    class="Preview"
  >
    <div
      class="firstLine"
    >
      <h2
        class="bp3-heading protocolName"
      >
        MVAC Intensifié 
      </h2>
      <div
        class="dayOneEquals"
      >
        J1 = J
        15
      </div>
    </div>
    <br />
    <p>
      Chimiothérapie comprenant 
      <strong />
      . réévaluée tous les 
      <strong>
        4
      </strong>
       
      cycles avec ...
    </p>
    <br />
    <div
      class="evaluations"
    >
      <h5
        class="bp3-heading"
      >
        Réévaluations : 
      </h5>
      <div
        class="evaluationBox"
      >
        <div
          class="flexLine"
        >
          <div
            class="evaluation"
          >
            4
             cycles
          </div>
          <div
            class="evaluationContent"
          >
            <strong>
              Consultation
            </strong>
            <span>
               et 
              <strong>
                Imagerie
              </strong>
            </span>
          </div>
        </div>
      </div>
      <br />
      <br />
      <h5
        class="bp3-heading"
      >
        Produits : 
      </h5>
      <div
        class="productLines"
      >
        <div
          class="productLine"
        >
          <strong>
            Vinblastine
          </strong>
          <br />
          <span
            class="delivered"
          >
            J
            1
          </span>
          <span
            class="notDelivered"
          >
            J
            2
          </span>
          <span
            class="notDelivered"
          >
            J
            3
          </span>
          <span
            class="notDelivered"
          >
            J
            4
          </span>
          <span
            class="notDelivered"
          >
            J
            5
          </span>
          <span
            class="notDelivered"
          >
            J
            6
          </span>
          <span
            class="notDelivered"
          >
            J
            7
          </span>
          <span
            class="notDelivered"
          >
            J
            8
          </span>
          <span
            class="notDelivered"
          >
            J
            9
          </span>
          <span
            class="notDelivered"
          >
            J
            10
          </span>
          <span
            class="notDelivered"
          >
            J
            11
          </span>
          <span
            class="notDelivered"
          >
            J
            12
          </span>
          <span
            class="notDelivered"
          >
            J
            13
          </span>
          <span
            class="notDelivered"
          >
            J
            14
          </span>
          <span
            class="notDelivered"
          >
            J
            15
          </span>
        </div>
        <div
          class="productLine"
        >
          <strong>
            Méthotrexate
          </strong>
          <br />
          <span
            class="delivered"
          >
            J
            1
          </span>
          <span
            class="notDelivered"
          >
            J
            2
          </span>
          <span
            class="notDelivered"
          >
            J
            3
          </span>
          <span
            class="notDelivered"
          >
            J
            4
          </span>
          <span
            class="notDelivered"
          >
            J
            5
          </span>
          <span
            class="notDelivered"
          >
            J
            6
          </span>
          <span
            class="notDelivered"
          >
            J
            7
          </span>
          <span
            class="notDelivered"
          >
            J
            8
          </span>
          <span
            class="notDelivered"
          >
            J
            9
          </span>
          <span
            class="notDelivered"
          >
            J
            10
          </span>
          <span
            class="notDelivered"
          >
            J
            11
          </span>
          <span
            class="notDelivered"
          >
            J
            12
          </span>
          <span
            class="notDelivered"
          >
            J
            13
          </span>
          <span
            class="notDelivered"
          >
            J
            14
          </span>
          <span
            class="notDelivered"
          >
            J
            15
          </span>
        </div>
        <div
          class="productLine"
        >
          <strong>
            Cisplatine
          </strong>
          <br />
          <span
            class="notDelivered"
          >
            J
            1
          </span>
          <span
            class="delivered"
          >
            J
            2
          </span>
          <span
            class="notDelivered"
          >
            J
            3
          </span>
          <span
            class="notDelivered"
          >
            J
            4
          </span>
          <span
            class="notDelivered"
          >
            J
            5
          </span>
          <span
            class="notDelivered"
          >
            J
            6
          </span>
          <span
            class="notDelivered"
          >
            J
            7
          </span>
          <span
            class="notDelivered"
          >
            J
            8
          </span>
          <span
            class="notDelivered"
          >
            J
            9
          </span>
          <span
            class="notDelivered"
          >
            J
            10
          </span>
          <span
            class="notDelivered"
          >
            J
            11
          </span>
          <span
            class="notDelivered"
          >
            J
            12
          </span>
          <span
            class="notDelivered"
          >
            J
            13
          </span>
          <span
            class="notDelivered"
          >
            J
            14
          </span>
          <span
            class="notDelivered"
          >
            J
            15
          </span>
        </div>
        <div
          class="productLine"
        >
          <strong>
            Doxorubicine
          </strong>
          <br />
          <span
            class="notDelivered"
          >
            J
            1
          </span>
          <span
            class="delivered"
          >
            J
            2
          </span>
          <span
            class="notDelivered"
          >
            J
            3
          </span>
          <span
            class="notDelivered"
          >
            J
            4
          </span>
          <span
            class="notDelivered"
          >
            J
            5
          </span>
          <span
            class="notDelivered"
          >
            J
            6
          </span>
          <span
            class="notDelivered"
          >
            J
            7
          </span>
          <span
            class="notDelivered"
          >
            J
            8
          </span>
          <span
            class="notDelivered"
          >
            J
            9
          </span>
          <span
            class="notDelivered"
          >
            J
            10
          </span>
          <span
            class="notDelivered"
          >
            J
            11
          </span>
          <span
            class="notDelivered"
          >
            J
            12
          </span>
          <span
            class="notDelivered"
          >
            J
            13
          </span>
          <span
            class="notDelivered"
          >
            J
            14
          </span>
          <span
            class="notDelivered"
          >
            J
            15
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
